### PR TITLE
Azure offloader

### PR DESCRIPTION
**Pull request by david-streamlio**
### Motivation

Added Azure offloader for Pulsar Tiered Storage

### Modifications

Refactored the tiered storage configuration tree to make it easier to add other JCloud Storage Providers in the future

### Result

Users will be able to use their Azure Storage accounts for tiered storage of Pulsar data.

### CODE REVIEW

**Comment by jiazhai** [comment]
@david-streamlio , Thanks for the work.  
There was my typo for "jcloud", of Since this PR brings in some refine work, Would you please also help change "jcloud" to "jclouds" in the file path and code?

**Comment by sijie** [comment]
> If possible, Would you please also help change "jcloud" to "jclouds" in the file path and code?

@jiazhai probably we can deal with this rename in a separate PR?

**Comment by david-streamlio** [comment]
I am in the process of updating the unit tests, as NONE of them actually used the constructor method that will be used in production.  I thought this would be a say change, but it turns out to be causing issues with the Mocks that wrap the Blobstore we are writing too.

`BlobStore spiedBlobStore = mock(BlobStore.class, delegatesTo(blobStore));`

it appears these mocks don't get invoked, resulting in test failures

**Comment by sijie** [comment]
Can we also update the doc page at https://pulsar.incubator.apache.org/docs/en/cookbooks-tiered-storage/?

**Comment by david-streamlio** [comment]
@sijie   Yes, I will update that document as well.

**Comment by david-streamlio** [comment]
Where is the content for the doc page at https://pulsar.incubator.apache.org/docs/en/cookbooks-tiered-storage/?

**Comment by sijie** [comment]
@david-streamlio the content is here : https://github.com/apache/pulsar/blob/master/site2/docs/cookbooks-tiered-storage.md

File: tiered-storage/jcloud/src/test/java/org/apache/bookkeeper/mledger/offload/jcloud/BlobStoreTestBase.java [hunk 1]
```diff
@@ -69,8 +88,9 @@ public void start() throws Exception {
     @AfterMethod
     public void tearDown() {
         if (blobStore != null &&
-            (!Boolean.parseBoolean(System.getProperty("testRealGCS", "false")) &&
-             !Boolean.parseBoolean(System.getProperty("testRealGCS", "false")))) {
+            (Boolean.parseBoolean(System.getProperty("testRealAWS", "false")) ||
```
Comment (jiazhai) [inline tiered-storage/jcloud/src/test/java/org/apache/bookkeeper/mledger/offload/jcloud/BlobStoreTestBase.java]: This change seems not right, for real test, we should not call delete container.

**Comment by jiazhai** [comment]
retest this please

File: tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jclouds/provider/factory/AWSBlobStoreFactory.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
+package org.apache.bookkeeper.mledger.offload.jclouds.provider.factory;
```
Comment (jiazhai) [inline tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jclouds/provider/factory/AWSBlobStoreFactory.java]: license header is missing.

**Comment by jiazhai** [comment]
@david-streamlio , Please fix the ut and header issues, Thanks.

File: tiered-storage/jcloud/src/test/java/org/apache/bookkeeper/mledger/offload/jcloud/BlobStoreTestBase.java [hunk 3]
```diff
@@ -88,9 +88,9 @@ public void start() throws Exception {
     @AfterMethod
     public void tearDown() {
         if (blobStore != null &&
-            (Boolean.parseBoolean(System.getProperty("testRealAWS", "false")) ||
-             Boolean.parseBoolean(System.getProperty("testRealGCS", "false")) || 
-             Boolean.parseBoolean(System.getProperty("testRealAzure", "false")))) {
+            (!Boolean.parseBoolean(System.getProperty("testRealAWS", "false")) ||
```
Comment (jiazhai) [inline tiered-storage/jcloud/src/test/java/org/apache/bookkeeper/mledger/offload/jcloud/BlobStoreTestBase.java]: "||" should be "&&" ?

**Comment by jiazhai** [comment]
rerun java8 tests

**Comment by jiazhai** [comment]
retest this please

**Comment by david-streamlio** [comment]
What is the expected behavior? So I can confirm the tests are correct.

**Comment by jiazhai** [comment]
rerun java8 tests

**Comment by jiazhai** [comment]
rerun java8 tests

**Comment by jiazhai** [comment]
rerun java8 tests

**Comment by david-streamlio** [comment]
I am open to suggestions on how to split these up, but these changes are tightly coupled. I could remove the Glacier offloader, but that would be just one class.

The refactor moved all of the logic for getting the credentials out of a single class with multiple if/the/else if/else if constructs and into a factory.  

Then I refactored the S3, and GCS to use the new factory methods, and added Azure.

**Comment by david-streamlio** [comment]
Following up on @ivankelly 's comment. I can submit a PR for the refactoring as it pertains to the AWS S3 offloader first, and then submit a different one for the Azure offloader once the first one is approved.  Will that be the best path forward @sijie ?

**Comment by sijie** [comment]
@david-streamlio let's get #2865 merged first. then you can rebase this pull request on #2865

**Comment by sijie** [comment]
Close this pull request as Azure Offloader has been merged.

**Pull request closed without merging**